server/src/main/java/org/opensearch/action/ActionRequest.java

```diff
@@ -58,6 +60,14 @@ public ActionRequest(StreamInput in) throws IOException {
         super(in);
     }
 
+    /**
+     * Should be overwritten by implementing plugins with the resources each actionRequest requires.
+     * Bypasses resource evaluation by default.
+     */
+    public List<String> getResources() {
```

Comment:
I think we have already touched upon resources: what are resources? how they are related to action? (and action request)?

Reply:
HI @reta, thanks for following up.

I am not sure what you mean by touched upon resources, but for context this is coming from: https://github.com/opensearch-project/security/issues/4500.

This framework modifies the ActionRequest with the override-able method so that a plugin can define a set of resources its actions interact with. The expectation is that plugins such as the Ml-Commons which uses model-groups would override this method so that the request can pass along that their action interacts on "model-group-1" for example.

From this information, you can then perform roles based access control or an an alternative form of authc/z on the request.

The change to the transportService is required so that we can define TransportRequests following the pattern /plugin/resource/add/<name> etc.

Reply:
> I am not sure what you mean by touched upon resources, but for context this is coming from: [opensearch-project/security#4500](https://github.com/opensearch-project/security/issues/4500).

I still don't understand this concept:
 - why resources are just arbitrary strings?
 - how the resources are being used?
 - what is the difference between resource and view?

Reply:
The ActionRequest class should not be modified. With ActionRequests that pertain to actions that involve indices, there is another interface called [IndicesRequest](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/IndicesRequest.java#L47) which they will implement to indicate that the request uses indices. Items within BulkRequests or MultiGet requests can be IndicesRequests.

If Resource Permissions are implemented as an abstraction of Index Permissions, then I think following a similar pattern with a new interface would make sense, but there is some discussion on the proposal about different approaches to providing a mechanism from the security plugin that plugins can utilize to protect "resources" created by the plugin: https://github.com/opensearch-project/security/issues/4500

Reply:
@cwperks please look at the discussion on the linked RFC @DarshitChanpura created. I understand you and he had some differences of opinion but the finalized design explicitly details modification of the ActionRequest as discussed by him and Nils. 

You have been advocating a re-design of the resource permissions to make them a form of index permission however this was never agreed upon on the RFC. To say that this class "should not  be modified" based on your preferred alternative is not a fair representation in my opinion.

Reply:
> You have been advocating a re-design of the resource permissions to make them a form of index permission however this was never agreed upon on the RFC. To say that this class "should not be modified" based on your preferred alternative is not a fair representation in my opinion.

The proposed design in the RFC is about making a `resource_permissions` section in a role definition where `resource_permissions` are an abstraction of `index_permissions` and the same glob matching done with index names would be applied to actions that apply to resources.

I wanted to leave a comment about how IndicesRequests worked because not all ActionRequests are IndicesRequests. The Security plugin has some pretty convoluted logic to resolve [concrete indices from a raw action request](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L616-L808). It would be nice if there were a better way to resolve indices from a raw action request. 

I left a [comment](https://github.com/opensearch-project/security/issues/4500#issuecomment-2233326609) about an idea of "resource sharing" which could provide a more backward compatible approach for ml commons. In [Model Access Control for ML Commons](https://github.com/opensearch-project/documentation-website/blob/main/_ml-commons-plugin/model-access-control.md#model-groups), there is a notion of `public`, `private` and `restricted` that specify how a model group can be shared. How would these translate to the new model for resource security?

Reply:
Thanks for following up @cwperks. 

I understand your concerns and am not saying that your preferred alternative may not be an advisable approach. 

My point is simply that the existing design makes it explicit how these changes should occur--and they would require modifying the ActionRequest specifically.

I will caution against the creation of any concrete Resource object inside OpenSearch. As you know, resolving indices is challenging to begin with and I don't know how easy it would be to resolve Resources as a concrete object across the cluster. 

For your concern on ML commons I would expect Roles based access control to be followed: https://github.com/opensearch-project/security/pull/4638#issuecomment-2286802299

Perhaps you think this would not work though? I cede to your judgement there, but that is my interpretation of the existing design intent. 

Just let me know what you think :)